### DISCUSSION TITLE
problem using symmetry

### DISCUSSION THREAD

**Discussion by 000sadaf** [QuSpin/QuSpin#493] [Q&A]
Hi,
i wanted to reduced the size of my hilbert space because when i ran codes greater than L = 12 it gives me ,Memory error.
i was using Nup = L//2.
i was finding the correlation function for SzSz , S+S-, S-S+ , SxSx, SySy, for the ground state. It gives me all the values for every correlation operator for GS, but problem is when i used symmetry Nup=L//2 
only SzSz correlation function gives me value but others correlations S+S-, S-S+, SxSx, SySy gives me ZERO.
WHY??????
import quspin as qp
from quspin.operators import hamiltonian
from quspin.basis import spin_basis_1d
import numpy as np
import sympy as sp
import matplotlib.pyplot as plt

L = 12
Jxy = 1.0  #spin interaction
Jzz = 1.0 # z interaction term

basis = spin_basis_1d(L,S="1/2",Nup=L//2,pauli=False)

jzz = [[Jzz, i, (i+1)%L]for i in range(L)]
jxy = [[Jxy/2, i, (i+1)%L]for i in range(L)]

static = [["+-", jxy], ["-+", jxy],["zz", jzz]]
dynamic = []
H_XXZ = hamiltonian(static, dynamic, basis=basis, dtype=np.float64)
E,V = H_XXZ.eigsh(k=1,which="SA")
psi_GS = V[:,0]
print(psi_GS)

``
Sz0_list = [["+",[0],1]]

Sz0_GS = basis.inplace_Op(psi_GS,Sz0_list,np.float64)

for r in range(L):
    #print(r)
    Szr_list = [["-",[r],1]]
    Szr_GS = basis.inplace_Op(psi_GS,Szr_list,np.float64,conjugated=True,transposed=True)
    corr = np.vdot(Sz0_GS,Szr_GS)
    print(corr)

**Comment by JiahaoYao** [comment]
Hi @000sadaf , my understanding of your issue is since you are restricting the basis to the `Nup=L//2` sectors, then if the wavefunction does not satisfies this symm, it will be zero. 

To be more specific, 

```python
Sz0_list = [["+",[0],1]]
Sz0_GS = basis.inplace_Op(psi_GS,Sz0_list,np.float64)
```

Here, when you compute this quantity, `phi_GS` satisfies the `Nup=L//2` symm, but when you apply `Sz0_list = [["+",[0],1]]`, then symm is no longer satisfies, the same thing applies for `X` (bit flip), `-`, and `Y`. However, the `Z` still preserves the symm, that is why `Sz0_GS` is not zero in that case.

**Comment by JiahaoYao** [comment]
I am not very familiar with the correlation's definition, so if you can define that again here, probably that would be perfect. 

Since it seems that when `S+S-`'s correlation is computed, the hamiltonian is separated. 


e.g. 
```python
Sz0_list = [["+",[0],1]]

Szr_list = [["-",[r],1]]
```

But if they can be combined together, the symm still satisfies, which i mean 

```python
Sz0_list = [["+-",[0],1]]

Sz0_GS = basis.inplace_Op(psi_GS,Sz0_list,np.float64)
corr = np.vdot(Sz0_GS,Szr_GS)
```

**Reply by 000sadaf** [reply]
Sz0_list = [["+-",[0],1]]

Sz0_GS = basis.inplace_Op(psi_GS,Sz0_list,np.float64)
corr = np.vdot(Sz0_GS,Szr_GS)

i cannot use this code because it means that S+(0)S-(0) only.
i made program for correlations S+(0)S-(r) , Sz(0)Sz(r) where r runs from1 to 12.


okay if S+S- is not using the symmetry, So my question is , how can i reduced my  Hilbert space (my Hamiltonian) so that i can run the program for higher spins like L=50 or just greater than 12.
my other question is why the psiGS change when using the Nup symmetry, PhiGS is supposed to be same because the ground state energy does not change also.

**Reply by JiahaoYao** [reply]
i can answer the last question, phi_GS changes because the basis is changed and ground state stays the same as long as the ground state satisfies the symm. 

```python
import quspin as qp
from quspin.operators import hamiltonian
from quspin.basis import spin_basis_1d
import numpy as np
import sympy as sp
import matplotlib.pyplot as plt

L = 4
Jxy = 1.0 #spin interaction
Jzz = 1.0 # z interaction term

basis = spin_basis_1d(L,S="1/2",pauli=False)


jzz = [[Jzz, i, (i+1)%L]for i in range(L)]
jxy = [[Jxy/2, i, (i+1)%L]for i in range(L)]


static = [["+-", jxy], ["-+", jxy],["zz", jzz]]
dynamic = []
H_XXZ = hamiltonian(static, dynamic, basis=basis, dtype=np.float64)
E,V = H_XXZ.eigsh(k=1,which="SA")
psi_GS = V[:,0]
print(E[0])


basis = spin_basis_1d(L,S="1/2",Nup=L//2,pauli=False)
H_XXZ = hamiltonian(static, dynamic, basis=basis, dtype=np.float64)
E,V = H_XXZ.eigsh(k=1,which="SA")
psi_GS = V[:,0]
print(E[0])
```
